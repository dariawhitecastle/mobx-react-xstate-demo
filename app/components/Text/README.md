<!-- ! This is a generated file. To make changes, edit <Component>.doc.js ! -->
## createText
Arbitrary text.

## Usage

```javascript
import { Text } from 'components/Text';
<Text />
```

## Properties

**a11yTitle**

Custom title to be used by screen readers.

```
string
```

**alignSelf**

How to align along the cross axis when contained in
      a Box or along the column axis when contained in a Grid.

```
start
center
end
stretch
```

**gridArea**

The name of the area to place
    this inside a parent Grid.

```
string
```

**margin**

The amount of margin around the component. An object can
      be specified to distinguish horizontal margin, vertical margin, and
      margin on a particular side.

```
none
xxsmall
xsmall
small
medium
large
xlarge
{
  bottom: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  horizontal: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  left: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  right: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  top: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string,
  vertical: 
    xxsmall
    xsmall
    small
    medium
    large
    xlarge
    string
}
string
```

**color**

A color identifier to use for the text color.

```
string
{
  dark: string,
  light: string
}
```

**size**

The font size and line height are primarily driven by the chosen tag. But, it can
be adjusted via this size property. The tag should be set for semantic
correctness and accessibility. This size property allows for stylistic
adjustments. Defaults to `medium`.

```
xsmall
small
medium
large
xlarge
xxlarge
string
```

**tag**

The DOM tag to use for the element. NOTE: This is deprecated in favor
         of indicating the DOM tag via the 'as' property.

```
string
function
```

**as**

The DOM tag or react component to use for the element. Defaults to `span`.

```
string
function
```

**textAlign**

How to align the text inside the component. Defaults to `start`.

```
start
center
end
```

**truncate**

Restrict the text to a single line and truncate with ellipsis if it
is too long to all fit.

```
boolean
```

**weight**

Font weight

```
normal
bold
number
```

**wordBreak**

Whether words should break when reaching the end of a line. Defaults to `normal`.

```
normal
break-all
keep-all
break-word
```
  
## Intrinsic element

```
span
```
## Theme
  
**global.colors.text**

The text color used for Text. Expects `object | { dark: string, light: string }`.

Defaults to

```
{ dark: '#f8f8f8', light: '#444444' }
```

**text**

The possible sizes of the text in terms of its font-size and line-height. Expects `object`.

Defaults to

```
{
      xsmall: {
        size: '12px',
        height: '18px',
       },
      small: {
        size: '14px',
        height: '20px',
       },
      medium: {
        size: '18px',
        height: '24px',
      },
      large: {
        size: '22px',
        height: '28px',
      },
      xlarge: {
        size: '26px',
        height: '32px',
      },
      xxlarge: {
        size: '34px',
        height: '40px',
      },
    }
```

**text.extend**

Any additional style for Text. Expects `string | (props) => {}`.

Defaults to

```
undefined
```

**global.edgeSize**

The possible sizes for margin. Expects `object`.

Defaults to

```
{
      edgeSize: {
        none: '0px',
        hair: '1px',
        xxsmall: '3px',
        xsmall: '6px',
        small: '12px',
        medium: '24px',
        large: '48px',
        xlarge: '96px',
        responsiveBreakpoint: 'small',
      },
    }
```
