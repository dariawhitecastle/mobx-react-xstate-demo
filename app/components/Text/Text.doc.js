import { describe, PropTypes } from 'react-desc'

import { colorPropType, genericProps } from '../../utils/propTypes'
import { themeDocUtils } from '../../utils/themeDocUtils'

export const doc = Text => {
  const DocumentedText = describe(Text)
    .description('Arbitrary text.')
    .usage(
      `import { Text } from 'components/Text';
<Text />`,
    )
    .intrinsicElement('span')

  DocumentedText.propTypes = {
    ...genericProps,
    color: colorPropType.description('A color identifier to use for the text color.'),
    size: PropTypes.oneOfType([
      PropTypes.oneOf(['xsmall', 'small', 'medium', 'large', 'xlarge', 'xxlarge']),
      PropTypes.string,
    ])
      .description(
        `The font size and line height are primarily driven by the chosen tag. But, it can
be adjusted via this size property. The tag should be set for semantic
correctness and accessibility. This size property allows for stylistic
adjustments.`,
      )
      .defaultValue('medium'),
    tag: PropTypes.oneOfType([PropTypes.string, PropTypes.func]).description(
      `The DOM tag to use for the element. NOTE: This is deprecated in favor
         of indicating the DOM tag via the 'as' property.`,
    ),
    as: PropTypes.oneOfType([PropTypes.string, PropTypes.func])
      .description(`The DOM tag or react component to use for the element.`)
      .defaultValue('span'),
    textAlign: PropTypes.oneOf(['start', 'center', 'end'])
      .description('How to align the text inside the component.')
      .defaultValue('start'),
    truncate: PropTypes.bool
      .description(
        `Restrict the text to a single line and truncate with ellipsis if it
is too long to all fit.`,
      )
      .defaultValue(false),
    weight: PropTypes.oneOfType([
      PropTypes.oneOf(['normal', 'bold']),
      PropTypes.number,
    ]).description('Font weight'),
    wordBreak: PropTypes.oneOf(['normal', 'break-all', 'keep-all', 'break-word'])
      .description('Whether words should break when reaching the end of a line.')
      .defaultValue('normal'),
  }

  return DocumentedText
}

export const themeDoc = {
  'global.colors.text': {
    description: 'The text color used for Text.',
    type: 'object | { dark: string, light: string }',
    defaultValue: "{ dark: '#f8f8f8', light: '#444444' }",
  },
  text: {
    description: `The possible sizes of the text in terms of its font-size and line-height.`,
    type: 'object',
    defaultValue: `{
      xsmall: {
        size: '12px',
        height: '18px',
       },
      small: {
        size: '14px',
        height: '20px',
       },
      medium: {
        size: '18px',
        height: '24px',
      },
      large: {
        size: '22px',
        height: '28px',
      },
      xlarge: {
        size: '26px',
        height: '32px',
      },
      xxlarge: {
        size: '34px',
        height: '40px',
      },
    }`,
  },
  'text.extend': {
    description: 'Any additional style for Text.',
    type: 'string | (props) => {}',
    defaultValue: undefined,
  },
  ...themeDocUtils.edgeStyle('The possible sizes for margin.'),
}
