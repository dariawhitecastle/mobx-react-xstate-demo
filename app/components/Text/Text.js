/* eslint-disable react/prop-types */

import React from 'react'
import { Text } from 'grommet'

import { doc } from './Text.doc'
import helpers from '../../utils/helpers'

/**
 *
 * Text
 *
 */
function createText({
  alignSelf,
  a11yTitle,
  as,
  gridArea,
  margin,
  children,
  color,
  level,
  responsive,
  size,
  textAlign,
  tag,
  truncate,
  weight,
  wordBreak,
  ...rest
}) {
  return (
    <Text
      alignSelf={alignSelf}
      a11yTitle={a11yTitle}
      as={as}
      gridArea={gridArea}
      margin={margin}
      color={color}
      level={level}
      responsive={responsive}
      size={size}
      textAlign={textAlign}
      tag={tag}
      truncate={truncate}
      weight={weight}
      wordBreak={wordBreak}
      {...rest}
    >
      {children}
    </Text>
  )
}

export default helpers.createWithDoc({
  docFunction: doc,
  component: createText,
})
