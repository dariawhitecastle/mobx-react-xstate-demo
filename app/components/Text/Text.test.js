import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import { Grommet } from 'grommet'
import Text from './Text'

test('renders', () => {
  const component = renderer.create(
    <Grommet>
      <Text>text</Text>
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders size', () => {
  const component = renderer.create(
    <Grommet>
      <Text size="xsmall" />
      <Text size="small" />
      <Text size="medium" />
      <Text size="large" />
      <Text size="xlarge" />
      <Text size="xxlarge" />
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders textAlign', () => {
  const component = renderer.create(
    <Grommet>
      <Text textAlign="start" />
      <Text textAlign="center" />
      <Text textAlign="end" />
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders margin', () => {
  const component = renderer.create(
    <Grommet>
      <Text margin="small" />
      <Text margin="medium" />
      <Text margin="large" />
      <Text margin="none" />
      <Text margin={{ vertical: 'small' }} />
      <Text margin={{ horizontal: 'small' }} />
      <Text margin={{ bottom: 'small' }} />
      <Text margin={{ top: 'small' }} />
      <Text margin={{ left: 'small' }} />
      <Text margin={{ right: 'small' }} />
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

const LONG = 'a b c d e f g h i j k l m n o p q r s t u v w x y z'

test('renders truncate', () => {
  const component = renderer.create(
    <Grommet>
      <Text truncate={false}>{LONG}</Text>
      <Text truncate>{LONG}</Text>
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders color', () => {
  const component = renderer.create(
    <Grommet>
      <Text color="status-critical" />
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('renders tag', () => {
  const component = renderer.create(
    <Grommet>
      <Text as="div" />
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

test('proxies tag', () => {
  const tagComponent = renderer.create(
    <Grommet>
      <Text tag="div" />
    </Grommet>,
  )
  const asComponent = renderer.create(
    <Grommet>
      <Text as="div" />
    </Grommet>,
  )
  expect(tagComponent.toJSON()).toEqual(asComponent.toJSON())
})

test('renders weight', () => {
  const component = renderer.create(
    <Grommet>
      <Text weight="normal" />
      <Text weight="bold" />
    </Grommet>,
  )
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
