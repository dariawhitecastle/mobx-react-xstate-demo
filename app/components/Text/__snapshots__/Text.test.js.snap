// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
.c1 {
  font-size: 18px;
  line-height: 24px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <span
    className="c1"
  >
    text
  </span>
</div>
`;

exports[`renders color 1`] = `
.c1 {
  font-size: 18px;
  line-height: 24px;
  color: #FF4040;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <span
    className="c1"
  />
</div>
`;

exports[`renders margin 1`] = `
.c1 {
  margin: 12px;
  font-size: 18px;
  line-height: 24px;
}

.c2 {
  margin: 24px;
  font-size: 18px;
  line-height: 24px;
}

.c3 {
  margin: 48px;
  font-size: 18px;
  line-height: 24px;
}

.c4 {
  margin: 0px;
  font-size: 18px;
  line-height: 24px;
}

.c5 {
  margin-top: 12px;
  margin-bottom: 12px;
  font-size: 18px;
  line-height: 24px;
}

.c6 {
  margin-left: 12px;
  margin-right: 12px;
  font-size: 18px;
  line-height: 24px;
}

.c7 {
  margin-bottom: 12px;
  font-size: 18px;
  line-height: 24px;
}

.c8 {
  margin-top: 12px;
  font-size: 18px;
  line-height: 24px;
}

.c9 {
  margin-left: 12px;
  font-size: 18px;
  line-height: 24px;
}

.c10 {
  margin-right: 12px;
  font-size: 18px;
  line-height: 24px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <span
    className="c1"
  />
  <span
    className="c2"
  />
  <span
    className="c3"
  />
  <span
    className="c4"
  />
  <span
    className="c5"
  />
  <span
    className="c6"
  />
  <span
    className="c7"
  />
  <span
    className="c8"
  />
  <span
    className="c9"
  />
  <span
    className="c10"
  />
</div>
`;

exports[`renders size 1`] = `
.c3 {
  font-size: 18px;
  line-height: 24px;
}

.c1 {
  font-size: 12px;
  line-height: 18px;
}

.c2 {
  font-size: 14px;
  line-height: 20px;
}

.c4 {
  font-size: 22px;
  line-height: 28px;
}

.c5 {
  font-size: 26px;
  line-height: 32px;
}

.c6 {
  font-size: 34px;
  line-height: 40px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <span
    className="c1"
    size="xsmall"
  />
  <span
    className="c2"
    size="small"
  />
  <span
    className="c3"
    size="medium"
  />
  <span
    className="c4"
    size="large"
  />
  <span
    className="c5"
    size="xlarge"
  />
  <span
    className="c6"
    size="xxlarge"
  />
</div>
`;

exports[`renders tag 1`] = `
.c1 {
  font-size: 18px;
  line-height: 24px;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <div
    className="c1"
  />
</div>
`;

exports[`renders textAlign 1`] = `
.c1 {
  font-size: 18px;
  line-height: 24px;
  text-align: left;
}

.c2 {
  font-size: 18px;
  line-height: 24px;
  text-align: center;
}

.c3 {
  font-size: 18px;
  line-height: 24px;
  text-align: right;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <span
    className="c1"
  />
  <span
    className="c2"
  />
  <span
    className="c3"
  />
</div>
`;

exports[`renders truncate 1`] = `
.c1 {
  font-size: 18px;
  line-height: 24px;
}

.c2 {
  font-size: 18px;
  line-height: 24px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <span
    className="c1"
  >
    a b c d e f g h i j k l m n o p q r s t u v w x y z
  </span>
  <span
    className="c2"
  >
    a b c d e f g h i j k l m n o p q r s t u v w x y z
  </span>
</div>
`;

exports[`renders weight 1`] = `
.c1 {
  font-size: 18px;
  line-height: 24px;
  font-weight: normal;
}

.c2 {
  font-size: 18px;
  line-height: 24px;
  font-weight: bold;
}

.c0 {
  font-size: 18px;
  line-height: 24px;
  box-sizing: border-box;
  -webkit-text-size-adjust: 100%;
  -ms-text-size-adjust: 100%;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
}

<div
  className="c0"
>
  <span
    className="c1"
  />
  <span
    className="c2"
  />
</div>
`;
